File: evidence/v1_1/ai_eval/subjects/counted/r1/chunk_sums/cpp/main.cpp
#include <cstdint>
#include <iostream>
#include <thread>
#include <vector>

struct ChunkResult {
    int64_t sum = 0;
    bool overflow = false;
};

static ChunkResult sum_range(const std::vector<int64_t>& values, size_t begin, size_t end) {
    ChunkResult result;
    for (size_t i = begin; i < end; ++i) {
        if (__builtin_add_overflow(result.sum, values[i], &result.sum)) {
            result.overflow = true;
            break;
        }
    }
    return result;
}

int main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int64_t k = 0, n = 0;
    std::cin >> k >> n;

    std::vector<int64_t> values(static_cast<size_t>(n));
    for (int64_t i = 0; i < n; ++i) {
        std::cin >> values[static_cast<size_t>(i)];
    }

    std::vector<ChunkResult> results(static_cast<size_t>(k));
    std::vector<std::thread> threads;
    threads.reserve(static_cast<size_t>(k));

    for (int64_t j = 0; j < k; ++j) {
        size_t begin = static_cast<size_t>(j * n / k);
        size_t end = static_cast<size_t>((j + 1) * n / k);
        threads.emplace_back([&results, &values, j, begin, end]() {
            results[static_cast<size_t>(j)] = sum_range(values, begin, end);
        });
    }

    for (auto& t : threads) {
        t.join();
    }

    bool total_overflow = false;
    int64_t total = 0;
    for (int64_t j = 0; j < k; ++j) {
        const ChunkResult& r = results[static_cast<size_t>(j)];
        if (r.overflow) {
            std::cout << "chunk " << j << " overflow\n";
            total_overflow = true;
        } else {
            std::cout << "chunk " << j << " " << r.sum << "\n";
            if (!total_overflow) {
                if (__builtin_add_overflow(total, r.sum, &total)) {
                    total_overflow = true;
                }
            }
        }
    }

    if (total_overflow) {
        std::cout << "total overflow\n";
    } else {
        std::cout << "total " << total << "\n";
    }

    return 0;
}

File: evidence/v1_1/ai_eval/subjects/counted/r1/chunk_sums/rust/main.rs
use std::io::{self, Read};
use std::thread;

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).expect("failed to read stdin");
    let mut it = input.split_ascii_whitespace();

    let k: usize = it.next().unwrap().parse().unwrap();
    let n: usize = it.next().unwrap().parse().unwrap();
    let values: Vec<i64> = (0..n).map(|_| it.next().unwrap().parse().unwrap()).collect();

    let mut bounds = Vec::with_capacity(k + 1);
    for j in 0..=k {
        bounds.push((j * n) / k);
    }

    let results: Vec<Option<i64>> = thread::scope(|scope| {
        let mut handles = Vec::with_capacity(k);
        for j in 0..k {
            let slice = &values[bounds[j]..bounds[j + 1]];
            handles.push(scope.spawn(move || {
                let mut sum: i64 = 0;
                for &v in slice {
                    match sum.checked_add(v) {
                        Some(s) => sum = s,
                        None => return None,
                    }
                }
                Some(sum)
            }));
        }
        handles.into_iter().map(|h| h.join().unwrap()).collect()
    });

    let mut total_overflow = false;
    let mut total: i64 = 0;
    for (j, r) in results.iter().enumerate() {
        match r {
            Some(s) => println!("chunk {} {}", j, s),
            None => {
                println!("chunk {} overflow", j);
                total_overflow = true;
            }
        }
    }

    if !total_overflow {
        for r in &results {
            let s = r.unwrap();
            match total.checked_add(s) {
                Some(t) => total = t,
                None => {
                    total_overflow = true;
                    break;
                }
            }
        }
    }

    if total_overflow {
        println!("total overflow");
    } else {
        println!("total {}", total);
    }
}
